src/Data/TaskMap.hs

```diff
@@ -88,14 +96,14 @@ memo
     -> (k -> IO v)
         -- ^ an action that is used to produce the value if the key isn't in the map.
     -> IO v
-memo tm@(TaskMap var) k task = do
-    -- TODO: should we insert another lookup here? It depends on how optimistic
-    -- we are that the lookup is successful.
-
-    a <- modifyMVarMasked var $ \m -> case HM.lookup k m of
-        Nothing -> do
-            !a <- asyncWithUnmask $ \umask -> umask (task k) `finally` delete tm k
-            m' <- evaluate $ HM.insert k a m
-            return (m', a)
-        (Just !a) -> return (m, a)
-    wait a
+memo tm@(TaskMap var) k task = bracket query cancel wait
+  where
+    query = do
+        -- TODO: should we insert another lookup here? It depends on how optimistic
```

Comment:
If this needs to be done, let's add an Asana ticket as well. Otherwise, it will be absorbed into the TODO blackhole...

Reply:
This TODO as old as this module and survived the introduction and abandonment of previous Asana usages :-)

I understand this more as a code comment for future code maintainers. When they get directed here during profiling, debugging, or refactoring, they may reconsider this. Since this is an open source project, it's not clear whether they'll have access to the currently used Asana instance.

We could changed it into "NOTE TO FUTURE MAINTAINERS", but that's a big lengthy. Maybe just "NOTE"?